### ISSUE TITLE
New warnings/errors?

### ISSUE THREAD

**Issue by Daniel-Cortez** [pawn-lang/compiler#528]
### Issue description:
Pawn is positioned as ["a robust language with a compiler that performs a maximum of static checks"](https://www.compuphase.com/pawn/pawn.htm#INTRODUCTION), but there are still a lot of areas where these checks could be improved.
I think it could be useful to have the following warnings/errors:
* **possible use of symbol before initialization**
  This one was already planned by the original author of Pawn as "warning 210" at some point of development (and even documented in the Language Manual), but for some reason the logic for this warning was never implemented. Experimental implementation is available in #440.

* **redundant code: switch argument is a constant value**
  We already have similar warnings for `if` statements:
```Pawn
if (false) {} // warning 205: redundant code: constant expression is zero
if (true) {} // warning 206: redundant code: constant expression is non-zero
```
  so why not adopt the same for `switch`?

* **negative or too big shift count**
  This can be achieved if the user has too many bit flags.

* **not all enum items are covered by switch statement**
  Actually, I already experimented with implementing this warning locally last year, and had a few false-positives in [YSI](https://github.com/pawn-lang/YSI-Includes/blob/5.x/YSI_Coding/y_hooks/y_hooks_impl.inc#L467) and amx_assembly ([1](https://github.com/Zeex/amx_assembly/blob/master/codescan.inc#L426), [2](https://github.com/Zeex/amx_assembly/blob/master/opcode.inc#L270)). I think restricting this warning to cases where only 1 or 2 enum items are missing should help to get rid from this kind of false-positives.

* **division by zero**
  Interestingly, the compiler already has a check for zero division in function `flooreddiv()` (sc3.c), but it only works if both operands are constant values:
```Pawn
new x = 1 / 0; // error 029: invalid expression, assumed zero
new y = x / 0; // no errors/warnings if the dividend is non-constant
```
  Also, the message "`invalid expression, assumed zero`" in the above example is misleading, because
  1) this is not the same as "division by zero";
  2) usually the compiler prints this message because of invalid syntax, while zero division is rather a semantic error.

### Workspace Information:

* Compiler version:
* Command line arguments provided (or sampctl version):
* Operating System:

**Comment by Daniel-Cortez** [comment]
Should I implement those warnings exactly as described? Would be nice to have any feedback before I get to implementing them.

**Comment by Daniel-Cortez** [comment]
Update: Added an idea for a new warning:
> * **enum increment "%s" has no effect on zero value**
> Might be useful if the user defined bit flags, but forgot to specify the value for the first enum item (or specified `0` instead of `1` by mistake):
> ```Pawn
> enum ePlayerFlags (<<= 1)
> {
>     pfEmailConfirmed,    // 0
>     pfTutorialCompleted, // enum increment "<<= 1" has no effect on zero value
> };
> ```
Any feedback is welcome.

**Comment by Daniel-Cortez** [comment]
Added ideas for two new diagnostics:

> * **variable "%s" used in loop condition not modified in loop body**
>   Detects situations when the user forgot to add an increment for the loop counter variable, thus making the loop endless:
> 
> ```pawn
> new index = 0;
> while (index < 10) // warning 2XX: variable "index" used in loop condition not modified in loop body
>     printf("%d", index);
> ```
> 
> * **none of the variables used in loop condition are modified in loop body**
>   Mostly the same as the previous diagnostic, except that it works when multiple variables are used in the loop condition.
> 
> ```pawn
> new i = 0, m = 10;
> while (i < m) // warning 2XX: none of the variables used in loop condition are modified in loop body
>     printf("%d", i);
> ```
> 
> The problem is that usually in such situations only one of the variables needs to be modified, but the compiler can't know which one.
> This new warning solves the issue by only telling that none of the variables were modified and allowing the user to decide which variable should be modified and which shouldn't.
